server/datastore/mysql/migrations/tables/20230315104937_EnsureUniformCollation.go

```diff
@@ -0,0 +1,73 @@
+package tables
+
+import (
+	"database/sql"
+	"fmt"
+)
+
+func init() {
+	MigrationClient.AddMigration(Up_20230315104937, Down_20230315104937)
+}
+
+// changeCollation changes the default collation set of the database and all
+// table to the provided collation
+//
+// This is based on the changeCharacterSet function that's included in this
+// module and part of the 20170306075207_UseUTF8MB migration.
+func changeCollation(tx *sql.Tx, charset string, collation string) error {
+	_, err := tx.Exec(fmt.Sprintf("ALTER DATABASE DEFAULT CHARACTER SET %s COLLATE %s", charset, collation))
+	if err != nil {
+		return fmt.Errorf("alter database: %w", err)
+	}
+
+	rows, err := tx.Query(`
+          SELECT table_name
+          FROM information_schema.TABLES AS T, information_schema.COLLATION_CHARACTER_SET_APPLICABILITY AS C
+          WHERE C.collation_name = T.table_collation
+          AND T.table_schema = (SELECT database())
+          AND (C.CHARACTER_SET_NAME != ? OR C.COLLATION_NAME != ?)`, charset, collation)
+	if err != nil {
+		return fmt.Errorf("selecting tables: %w", err)
+	}
+	defer rows.Close()
+
+	var names []string
+	for rows.Next() {
+		var name string
+		if err := rows.Scan(&name); err != nil {
+			return fmt.Errorf("scanning ID: %w", err)
+		}
+		names = append(names, name)
+	}
+
+	if err := rows.Err(); err != nil {
+		return fmt.Errorf("scanning rows: %w", err)
+	}
+
+	// disable foreign checks before changing the collations, otherwise the
+	// migration might fail. These are re-enabled afterwards.
+	if _, err := tx.Exec("SET FOREIGN_KEY_CHECKS = 0"); err != nil {
+		return fmt.Errorf("disabling foreign key checks: %w", err)
+	}
+	for _, name := range names {
+		_, err = tx.Exec(fmt.Sprintf("ALTER TABLE %s CONVERT TO CHARACTER SET %s COLLATE %s", name, charset, collation))
+		if err != nil {
+			return fmt.Errorf("alter table %s: %w", name, err)
+		}
+	}
+	if _, err := tx.Exec("SET FOREIGN_KEY_CHECKS = 1"); err != nil {
+		return fmt.Errorf("re-enabling foreign key checks: %w", err)
+	}
+	return nil
+}
+
+func Up_20230315104937(tx *sql.Tx) error {
+	// while newer versions of MySQL default to
+	// utf8mb4_0900_ai_ci, we still need to support 5.7, which
+	// defaults to utf8mb4_general_ci
```

Comment:
I wonder if there is reason to prefer one over the other?

Reply:
that's a good question, and there might be even a better option.

I do know that `utf8mb4_0900_ai_ci` is not supported by 5.7, so we can't use that, but maybe something else.

Reply:
It looks like this change was made because `utf8mb4_0900_ai_ci` provides better unicode support: https://dev.mysql.com/blog-archive/mysql-8-0-collations-migrating-from-older-collations/

Reply:
here are the ones we have available to us:

```mysql> SELECT VERSION();
+------------+
| VERSION()  |
+------------+
| 5.7.40-log |
+------------+
1 row in set (0.00 sec)

mysql> SHOW COLLATION WHERE charset = 'utf8mb4';
+------------------------+---------+-----+---------+----------+---------+
| Collation              | Charset | Id  | Default | Compiled | Sortlen |
+------------------------+---------+-----+---------+----------+---------+
| utf8mb4_general_ci     | utf8mb4 |  45 | Yes     | Yes      |       1 |
| utf8mb4_bin            | utf8mb4 |  46 |         | Yes      |       1 |
| utf8mb4_unicode_ci     | utf8mb4 | 224 |         | Yes      |       8 |
| utf8mb4_icelandic_ci   | utf8mb4 | 225 |         | Yes      |       8 |
| utf8mb4_latvian_ci     | utf8mb4 | 226 |         | Yes      |       8 |
| utf8mb4_romanian_ci    | utf8mb4 | 227 |         | Yes      |       8 |
| utf8mb4_slovenian_ci   | utf8mb4 | 228 |         | Yes      |       8 |
| utf8mb4_polish_ci      | utf8mb4 | 229 |         | Yes      |       8 |
| utf8mb4_estonian_ci    | utf8mb4 | 230 |         | Yes      |       8 |
| utf8mb4_spanish_ci     | utf8mb4 | 231 |         | Yes      |       8 |
| utf8mb4_swedish_ci     | utf8mb4 | 232 |         | Yes      |       8 |
| utf8mb4_turkish_ci     | utf8mb4 | 233 |         | Yes      |       8 |
| utf8mb4_czech_ci       | utf8mb4 | 234 |         | Yes      |       8 |
| utf8mb4_danish_ci      | utf8mb4 | 235 |         | Yes      |       8 |
| utf8mb4_lithuanian_ci  | utf8mb4 | 236 |         | Yes      |       8 |
| utf8mb4_slovak_ci      | utf8mb4 | 237 |         | Yes      |       8 |
| utf8mb4_spanish2_ci    | utf8mb4 | 238 |         | Yes      |       8 |
| utf8mb4_roman_ci       | utf8mb4 | 239 |         | Yes      |       8 |
| utf8mb4_persian_ci     | utf8mb4 | 240 |         | Yes      |       8 |
| utf8mb4_esperanto_ci   | utf8mb4 | 241 |         | Yes      |       8 |
| utf8mb4_hungarian_ci   | utf8mb4 | 242 |         | Yes      |       8 |
| utf8mb4_sinhala_ci     | utf8mb4 | 243 |         | Yes      |       8 |
| utf8mb4_german2_ci     | utf8mb4 | 244 |         | Yes      |       8 |
| utf8mb4_croatian_ci    | utf8mb4 | 245 |         | Yes      |       8 |
| utf8mb4_unicode_520_ci | utf8mb4 | 246 |         | Yes      |       8 |
| utf8mb4_vietnamese_ci  | utf8mb4 | 247 |         | Yes      |       8 |
+------------------------+---------+-----+---------+----------+---------+
26 rows in set (0.02 sec)
```

at first glance `utf8mb4_general_ci`, `utf8mb4_unicode_ci`, `utf8mb4_bin` and `utf8mb4_unicode_520_ci` have non-specific names, I have no idea which one is best.

my instinct is to keep it at `utf8mb4_general_ci` which was the default.

Reply:
Agreed on keeping it with `utf8mb4_general_ci`.

Reply:
Seems like a pretty clear reason to prefer that one :)

FWIW I do think we should drop MySQL 5.7 support sometime soon-ish, but now is not the time.

Reply:
> FWIW I do think we should drop MySQL 5.7 support sometime soon-ish, but now is not the time.

yess!! can't wait!

Reply:
Agreed `utf8mb4_unicode_ci` seems like the way to go.

Reply:
found more info in the MySQL docs: https://dev.mysql.com/doc/refman/8.0/en/charset-unicode-sets.html#charset-unicode-sets-general-versus-unicode


> MySQL implements language-specific Unicode collations if the ordering with utf8mb4_unicode_ci does not work well for a language. For example, utf8mb4_unicode_ci works fine for German dictionary order and French, so there is no need to create special utf8mb4 collations.
> 
> utf8mb4_general_ci also is satisfactory for both German and French, except that ß is equal to s, and not to ss. If this is acceptable for your application, you should use utf8mb4_general_ci because it is faster. If this is not acceptable (for example, if you require German dictionary order), use utf8mb4_unicode_.ci because it is more accurate. 

I wish I knew more about this to make a more informed decision

Reply:
FWIW my understanding of the collation is that it's about how the values are treated for sorting and comparison. While we certainly see some unicode in Fleet's data, I don't think there is enough that users will notice small differences in sort orders for extended character sets.

Reply:
sounds good 👍! yes I'm just worried about doing this again in the future, it's scary and a bit painful